File: apps/go/chapter03/search_test.go
package chapter03_test

import (
	"testing"

	"github.com/k2works/getting-started-algorithm/go/chapter03"
)

func TestSsearchWhile(t *testing.T) {
	a := []int{6, 4, 3, 2, 1, 2, 8}
	got := chapter03.SsearchWhile(a, 2)
	if got != 3 {
		t.Errorf("SsearchWhile(..., 2) = %d; want 3", got)
	}
	got = chapter03.SsearchWhile(a, 99)
	if got != -1 {
		t.Errorf("SsearchWhile(..., 99) = %d; want -1", got)
	}
}

func TestSsearchFor(t *testing.T) {
	a := []int{6, 4, 3, 2, 1, 2, 8}
	got := chapter03.SsearchFor(a, 2)
	if got != 3 {
		t.Errorf("SsearchFor(..., 2) = %d; want 3", got)
	}
}

func TestBsearch(t *testing.T) {
	a := []int{1, 2, 3, 5, 7, 8, 9}
	got := chapter03.Bsearch(a, 5)
	if got != 3 {
		t.Errorf("Bsearch(..., 5) = %d; want 3", got)
	}
	got = chapter03.Bsearch(a, 99)
	if got != -1 {
		t.Errorf("Bsearch(..., 99) = %d; want -1", got)
	}
}

func TestSsearchSentinel(t *testing.T) {
	a := []int{6, 4, 3, 2, 1, 2, 8}
	got := chapter03.SsearchSentinel(a, 2)
	if got != 3 {
		t.Errorf("SsearchSentinel(..., 2) = %d; want 3", got)
	}
	got = chapter03.SsearchSentinel(a, 99)
	if got != -1 {
		t.Errorf("SsearchSentinel(..., 99) = %d; want -1", got)
	}
}

func TestHashTable(t *testing.T) {
	ht := chapter03.NewHashTable(10)
	ht.Add(1, "one")
	ht.Add(2, "two")
	ht.Add(3, "three")

	val, ok := ht.Search(2)
	if !ok || val != "two" {
		t.Errorf("HashTable.Search(2) = %v, %v; want two, true", val, ok)
	}

	_, ok = ht.Search(99)
	if ok {
		t.Error("HashTable.Search(99) should not be found")
	}

	ht.Remove(2)
	_, ok = ht.Search(2)
	if ok {
		t.Error("HashTable.Search(2) should not be found after Remove")
	}
}

func TestOpenHash(t *testing.T) {
	oh := chapter03.NewOpenHash(10)
	oh.Add(1, "one")
	oh.Add(2, "two")
	oh.Add(3, "three")

	val, ok := oh.Search(2)
	if !ok || val != "two" {
		t.Errorf("OpenHash.Search(2) = %v, %v; want two, true", val, ok)
	}

	_, ok = oh.Search(99)
	if ok {
		t.Error("OpenHash.Search(99) should not be found")
	}

	oh.Remove(2)
	_, ok = oh.Search(2)
	if ok {
		t.Error("OpenHash.Search(2) should not be found after Remove")
	}

	added := oh.Add(1, "one_dup")
	if added {
		t.Error("OpenHash.Add with duplicate key should return false")
	}
}

File: apps/c/chapter03/search_test.c
#include <stdio.h>
#include "../include/test_helper.h"
#include "search.h"

void test_ssearch_while(void) {
    int a[] = {6, 4, 3, 2, 1, 2, 8};
    ASSERT_EQ_INT(3, ssearch_while(a, 7, 2));
    ASSERT_EQ_INT(-1, ssearch_while(a, 7, 9));
}

void test_ssearch_for(void) {
    int a[] = {6, 4, 3, 2, 1, 2, 8};
    ASSERT_EQ_INT(3, ssearch_for(a, 7, 2));
    ASSERT_EQ_INT(-1, ssearch_for(a, 7, 9));
}

void test_ssearch_sentinel(void) {
    int a[] = {6, 4, 3, 2, 1, 2, 8};
    ASSERT_EQ_INT(3, ssearch_sentinel(a, 7, 2));
    ASSERT_EQ_INT(-1, ssearch_sentinel(a, 7, 9));
}

void test_bsearch_int(void) {
    int a[] = {1, 2, 3, 5, 7, 8, 9};
    ASSERT_EQ_INT(3, bsearch_int(a, 7, 5));
    ASSERT_EQ_INT(-1, bsearch_int(a, 7, 4));
}

void test_chained_hash(void) {
    ChainedHash *h = chained_hash_new(13);
    ASSERT_TRUE(chained_hash_add(h, 1, 100));
    ASSERT_TRUE(chained_hash_add(h, 14, 200)); /* same bucket as 1 */
    ASSERT_EQ_INT(100, chained_hash_search(h, 1));
    ASSERT_EQ_INT(200, chained_hash_search(h, 14));
    ASSERT_EQ_INT(-1,  chained_hash_search(h, 99));
    ASSERT_TRUE(chained_hash_remove(h, 1));
    ASSERT_EQ_INT(-1,  chained_hash_search(h, 1));
    chained_hash_free(h);
}

void test_open_hash(void) {
    OpenHash *h = open_hash_new(13);
    ASSERT_TRUE(open_hash_add(h, 1, 100));
    ASSERT_TRUE(open_hash_add(h, 14, 200));
    ASSERT_EQ_INT(100, open_hash_search(h, 1));
    ASSERT_EQ_INT(200, open_hash_search(h, 14));
    ASSERT_EQ_INT(-1,  open_hash_search(h, 99));
    ASSERT_TRUE(open_hash_remove(h, 1));
    ASSERT_EQ_INT(-1,  open_hash_search(h, 1));
    open_hash_free(h);
}

int main(void) {
    test_ssearch_while();
    test_ssearch_for();
    test_ssearch_sentinel();
    test_bsearch_int();
    test_chained_hash();
    test_open_hash();
    TEST_SUMMARY();
    return 0;
}
